sql/5_0_2-to-5_0_3_upgrade.sql

```diff
@@ -515,3 +515,82 @@ ALTER TABLE `api_token` ADD `token_api` varchar(40);
 #IfMissingColumn api_token patient_id
 ALTER TABLE `api_token` ADD `patient_id` bigint(20) NOT NULL;
 #EndIf
+
+#IfNotTable fhir_forced_id
+CREATE TABLE `fhir_forced_id` (
+    `pid` bigint(20) NOT NULL,
+    `forced_id` varchar(100) NOT NULL,
+    `resource_pid` bigint(20) NOT NULL,
+    `resource_type` varchar(100) DEFAULT '',
+    PRIMARY KEY (`pid`),
+    UNIQUE KEY `idx_forcedid_resid` (`resource_pid`),
+    UNIQUE KEY `idx_forcedid_type_resid` (`resource_type`,`resource_pid`),
+    KEY `idx_forcedid_type_forcedid` (`resource_type`,`forced_id`)
+) ENGINE=InnoDB;
+#EndIf
+
+#IfNotTable fhir_resource
+CREATE TABLE `fhir_resource` (
+    `res_id` bigint(20) NOT NULL,
+    `res_deleted_at` datetime DEFAULT NULL,
+    `res_version` varchar(7) DEFAULT NULL,
+    `has_tags` bit(1) NOT NULL,
+    `res_published` datetime DEFAULT NULL,
+    `res_updated` datetime DEFAULT NULL,
+    `sp_has_links` bit(1) DEFAULT NULL,
+    `hash_sha256` varchar(64) DEFAULT NULL,
+    `sp_index_status` bigint(20) DEFAULT NULL,
+    `res_language` varchar(20) DEFAULT NULL,
+    `sp_cmpstr_uniq_present` bit(1) DEFAULT NULL,
+    `sp_coords_present` bit(1) DEFAULT NULL,
+    `sp_date_present` bit(1) DEFAULT NULL,
+    `sp_number_present` bit(1) DEFAULT NULL,
+    `sp_quantity_present` bit(1) DEFAULT NULL,
+    `sp_string_present` bit(1) DEFAULT NULL,
+    `sp_token_present` bit(1) DEFAULT NULL,
+    `sp_uri_present` bit(1) DEFAULT NULL,
+    `res_profile` varchar(200) DEFAULT NULL,
+    `res_type` varchar(30) DEFAULT NULL,
+    `res_ver` bigint(20) DEFAULT NULL,
+    `forced_id_pid` bigint(20) DEFAULT NULL,
+    PRIMARY KEY (`res_id`),
+    KEY `idx_res_date` (`res_updated`),
+    KEY `idx_res_lang` (`res_type`,`res_language`),
+    KEY `idx_res_profile` (`res_profile`),
+    KEY `idx_res_type` (`res_type`),
+    KEY `idx_indexstatus` (`sp_index_status`),
+    KEY `fk_resource_forcedid` (`forced_id_pid`)
+) ENGINE=InnoDB;
+#EndIf
+
+#IfNotTable fhir_res_ver
+CREATE TABLE `fhir_res_ver` (
+    `pid` bigint(20) NOT NULL,
+    `res_deleted_at` datetime DEFAULT NULL,
+    `res_version` varchar(7) DEFAULT NULL,
+    `has_tags` bit(1) NOT NULL,
+    `res_published` datetime DEFAULT NULL,
+    `res_updated` datetime DEFAULT NULL,
+    `res_encoding` varchar(5) NOT NULL,
+    `res_text` longblob,
+    `res_id` bigint(20) DEFAULT NULL,
+    `res_type` varchar(30) NOT NULL,
+    `res_ver` bigint(20) NOT NULL,
+    `forced_id_pid` bigint(20) DEFAULT NULL,
+    PRIMARY KEY (`pid`),
+    UNIQUE KEY `idx_resver_id_ver` (`res_id`,`res_ver`),
+    KEY `idx_resver_type_date` (`res_type`,`res_updated`),
+    KEY `idx_resver_id_date` (`res_id`,`res_updated`),
+    KEY `idx_resver_date` (`res_updated`),
+    KEY `fk_resver_forcedid` (`forced_id_pid`)
+) ENGINE=InnoDB;
+
+ALTER TABLE `fhir_forced_id`
```

Comment:
I certainly agree with everyone that in 2020 foreign keys are not a bad thing :) Cascading constraints are also not a bad thing, but opinions may differ there. One thing to keep in mind is whether or not the database automatically indexes a foreign key constraint/column. It looks like [MySQL will auto-index a foreign key](https://dev.mysql.com/doc/refman/8.0/en/constraint-foreign-key.html), so I'm assuming Maria DB will as well :smile:

Reply:
I haven't seen an issue on Maria 10.1 which I run my HAPI against.
Still, worthy note and needs watching.
Brady does a good job at trying to keep us as compatible as possible against what environments we can run in but, I've got to tell you, this is getting harder by the day.
Between PHP versions, MySql/Maria and wayward browsers(harkens back to the old browser wars) at some point, we'll need to choose. IMO.

Reply:
@dixonwhitmire Yeah Right! I scarcely have the energy to decide if we want to support version much less do battle over a new anything. I'll be happy with using foreign keys and views next.:) 
Hey do you think it would be appropriate to use these tables for our standard API ?
I'm very wary of what gets committed to chart w/o some review and history?

Reply:
@sjpadgett it's never too late! I'll just leave [this](https://flask.palletsprojects.com/en/1.1.x/) right here :smile: But really other than an existing community developer base, current product support options, and the cost of rewriting, really what is there to stop you? 😆 

How do you think the tables would be used with the standard API? Would you give them each standard API resource a UUID/id?

Reply:
> Would you give them each standard API resource a UUID/id?

Sure and Brady is working hard on bringing in a mechanism do that. Thats next in our adventure.